event_handling/include/essentials/NotifyTimer.hpp

```diff
@@ -0,0 +1,180 @@
+    bool _started;                             /** < True, if the NotifyTimer is active. False, otherwise. */
+    std::chrono::milliseconds _msInterval;     /** < The milliseconds between two calls to the callback. */
+    std::chrono::milliseconds _msDelayedStart; /** < The milliseconds between (re)starting the NotifyTimer and the callback call. */
+    std::unique_ptr<std::thread> _runThread;
+};
+
+template <class NotificationClass>
+NotifyTimer<NotificationClass>::NotifyTimer(std::chrono::milliseconds msInterval, std::chrono::milliseconds msDelayedStart,
+        t_notificationcallback<NotificationClass> callback, NotificationClass* obj)
+        : _msInterval(msInterval)
+        , _msDelayedStart(msDelayedStart)
+        , _running(true)
+        , _started(false)
+        , _callback(callback)
+        , _obj(obj)
+{
+    /**
+     * We initialise the thread in the body of the constructor
+     * because this guarantees to have all control variables
+     * for this thread initialised properly before it starts.
+     */
+    _runThread = std::make_unique<std::thread>(&NotifyTimer::run, this, false);
+}
+
+template <class NotificationClass>
+NotifyTimer<NotificationClass>::NotifyTimer(std::chrono::milliseconds msInterval, std::chrono::milliseconds msDelayedStart, bool notifyAllThreads)
+        : _msInterval(msInterval)
+        , _msDelayedStart(msDelayedStart)
+        , _running(true)
+        , _started(false)
+        , _obj(nullptr)
+{
+    _runThread = std::make_unique<std::thread>(&NotifyTimer::run, this, notifyAllThreads);
+}
+
+template <class NotificationClass>
+void NotifyTimer<NotificationClass>::run(bool notifyAllThreads)
+{
+    while (_running) {
+        {
+            std::unique_lock<std::mutex> lck(_cv_mtx);
+            _cv.wait(lck, [&] { return !_running || _started; });
+        }
+
+        if (!_running) // for destroying the timer
+            return;
+
+        if (_msDelayedStart.count() > 0) {
+            std::this_thread::sleep_for(_msDelayedStart);
+        }
+
+        while (isStarted()) {
+            std::chrono::system_clock::time_point start = std::chrono::high_resolution_clock::now();
+            if (_obj != nullptr) {
+                (_obj->*_callback)();
+            } else {
+                notifyEveryCV(notifyAllThreads);
+            }
+            auto dura = std::chrono::high_resolution_clock::now() - start;
+            std::this_thread::sleep_for(_msInterval - dura);
+//            std::cout << "[NotifyTimer] Finished one iteration." << std::endl;
+        }
+    }
+}
+
+template <class NotificationClass>
+NotifyTimer<NotificationClass>::~NotifyTimer()
+{
+//    std::cout << "[NotifyTimer] DESTRUCT: _started: " << std::boolalpha << _started << " _running: " << _running << std::endl;
+    {
+        std::lock_guard<std::mutex> lockGuard(_cv_mtx);
+//        std::cout << "[NotifyTimer] DESTRUCT!" << std::endl;
```

Comment:
looks like the comments sneaked in by mistake

Reply:
done